File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Trpak/Archive/TrpakLoader.cs
using MiniToolbox.Core.Archive;
using System.Runtime.InteropServices;

namespace MiniToolbox.Trpak.Archive
{
    /// <summary>
    /// Loads TRPAK (GFLXPACK) archives.
    /// Ported from gftool GFPakSerializer.Deserialize().
    /// 
    /// Usage:
    ///   var cache = new TrpakHashCache();
    ///   cache.LoadHashList(File.ReadAllLines("hashlist.txt"));
    ///   var archive = TrpakLoader.Load("model.trpak", cache);
    ///   var trmdl = archive.FindFilesByExtension(".trmdl").First();
    /// </summary>
    public static class TrpakLoader
    {
        public static TrpakArchive Load(string path, TrpakHashCache? hashCache = null)
        {
            using var fs = File.OpenRead(path);
            using var br = new BinaryReader(fs);
            return Load(br, hashCache);
        }

        public static TrpakArchive Load(BinaryReader br, TrpakHashCache? hashCache = null)
        {
            hashCache ??= new TrpakHashCache();

            // Read header
            var header = ReadStruct<TrpakHeader>(br, TrpakHeader.Size);

            // Embedded file offset + hash offset
            ulong embeddedFileOff = br.ReadUInt64();
            ulong embeddedFileHashOff = br.ReadUInt64();

            // Folder offsets
            var folderOffsets = new long[header.FolderNumber];
            for (int i = 0; i < header.FolderNumber; i++)
                folderOffsets[i] = br.ReadInt64();

            // File hashes (absolute path hashes)
            br.BaseStream.Position = (long)embeddedFileHashOff;
            var fileHashes = new ulong[header.FileNumber];
            for (int i = 0; i < header.FileNumber; i++)
                fileHashes[i] = br.ReadUInt64();

            // File headers
            br.BaseStream.Position = (long)embeddedFileOff;
            var fileHeaders = new TrpakFileHeader[header.FileNumber];
            for (int i = 0; i < header.FileNumber; i++)
                fileHeaders[i] = ReadStruct<TrpakFileHeader>(br, TrpakFileHeader.Size);

            // Decompress files
            var files = new byte[header.FileNumber][];
            for (int i = 0; i < header.FileNumber; i++)
            {
                var fh = fileHeaders[i];
                var rawBytes = br.ReadBytes((int)fh.FileSize);

                files[i] = fh.CompressionType switch
                {
                    TrpakCompressionType.Lz4 => Lz4Decompressor.Decompress(rawBytes, (int)fh.BufferSize),
                    TrpakCompressionType.Oodle => OodleDecompressor.Decompress(rawBytes, (int)fh.BufferSize) ?? rawBytes,
                    _ => rawBytes
                };
            }

            // Build folder/file structure
            var archive = new TrpakArchive();
            for (int i = 0; i < header.FolderNumber; i++)
            {
                br.BaseStream.Position = folderOffsets[i];
                var folderHeader = ReadStruct<TrpakFolderHeader>(br, TrpakFolderHeader.Size);

                string folderName = hashCache.GetName(folderHeader.Hash) ?? folderHeader.Hash.ToString("X16");

                var folder = new TrpakFolder { Path = folderName };

                for (int j = 0; j < folderHeader.ContentNumber; j++)
                {
                    var content = ReadStruct<TrpakFolderIndex>(br, TrpakFolderIndex.Size);
                    int fileIndex = (int)content.Index;

                    string fileName = hashCache.GetName(content.Hash) ?? content.Hash.ToString("X16");

                    // Try to resolve the full path from the file hash
                    string fullPath = hashCache.GetName(fileHashes[fileIndex])
                                      ?? (folderName + fileName);

                    // Auto-register resolved paths
                    if (hashCache.GetName(fileHashes[fileIndex]) == null)
                        hashCache.Add(fileHashes[fileIndex], fullPath);

                    folder.Files.Add(new TrpakFile
                    {
                        Name = fileName,
                        FullName = fullPath,
                        Data = fileIndex < files.Length ? files[fileIndex] : Array.Empty<byte>()
                    });
                }

                archive.Folders.Add(folder);
            }

            return archive;
        }

        private static T ReadStruct<T>(BinaryReader br, int size) where T : struct
        {
            var bytes = br.ReadBytes(size);
            var handle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
            try
            {
                return Marshal.PtrToStructure<T>(handle.AddrOfPinnedObject());
            }
            finally
            {
                handle.Free();
            }
        }
    }
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/lib/Archive/TrpakLoader.ts
/**
 * TRPAK Archive Loader
 * Ported from C# TrpakLoader.cs
 * 
 * Usage:
 *   const cache = new TrpakHashCache();
 *   cache.LoadHashList(fs.readFileSync('hashlist.txt', 'utf8').split('\n'));
 *   const archive = TrpakLoader.Load('model.trpak', cache);
 *   const trmdl = archive.FindFilesByExtension('.trmdl')[0];
 */

import * as fs from 'fs';
import {
    TrpakHeader,
    TrpakFolderHeader,
    TrpakFolderIndex,
    TrpakFileHeader,
    TrpakCompressionType,
    TrpakArchive,
    TrpakFolder,
    TrpakFile,
    TrpakHashCache
} from './TrpakTypes.js';
import { OodleDecompressor, Lz4Decompressor } from './Decompressors.js';

export class TrpakLoader {
    static Load(path: string, hashCache?: TrpakHashCache): TrpakArchive {
        const data = fs.readFileSync(path);
        return this.LoadFromBuffer(data, hashCache);
    }

    static LoadFromBuffer(data: Buffer, hashCache?: TrpakHashCache): TrpakArchive {
        hashCache ??= new TrpakHashCache();

        let offset = 0;

        // Read header
        const header = this.ReadStructTrpakHeader(data, offset);
        offset = TrpakHeader.Size;

        // Embedded file offset + hash offset
        const embeddedFileOff = data.readBigUInt64LE(offset);
        offset += 8;
        const embeddedFileHashOff = data.readBigUInt64LE(offset);
        offset += 8;

        // Folder offsets
        const folderOffsets: number[] = [];
        for (let i = 0; i < header.FolderNumber; i++) {
            folderOffsets.push(Number(data.readBigInt64LE(offset)));
            offset += 8;
        }

        // File hashes (absolute path hashes)
        offset = Number(embeddedFileHashOff);
        const fileHashes: bigint[] = [];
        for (let i = 0; i < header.FileNumber; i++) {
            fileHashes.push(data.readBigUInt64LE(offset));
            offset += 8;
        }

        // File headers
        offset = Number(embeddedFileOff);
        const fileHeaders: TrpakFileHeader[] = [];
        for (let i = 0; i < header.FileNumber; i++) {
            fileHeaders.push(this.ReadStructTrpakFileHeader(data, offset));
            offset += TrpakFileHeader.Size;
        }

        // Decompress files
        const files: Buffer[] = [];
        for (let i = 0; i < header.FileNumber; i++) {
            const fh = fileHeaders[i];
            const rawBytes = data.subarray(offset, offset + fh.FileSize);
            offset += fh.FileSize;

            let decompressed: Buffer;
            switch (fh.CompressionType) {
                case TrpakCompressionType.Lz4:
                    decompressed = Lz4Decompressor.Decompress(rawBytes, fh.BufferSize);
                    break;
                case TrpakCompressionType.Oodle:
                    const oodleResult = OodleDecompressor.Decompress(rawBytes, fh.BufferSize);
                    decompressed = oodleResult ?? rawBytes;
                    break;
                default:
                    decompressed = Buffer.from(rawBytes);
            }
            files.push(decompressed);
        }

        // Build folder/file structure
        const archive = new TrpakArchive();
        for (let i = 0; i < header.FolderNumber; i++) {
            offset = folderOffsets[i];
            const folderHeader = this.ReadStructTrpakFolderHeader(data, offset);
            offset += TrpakFolderHeader.Size;

            const folderName = hashCache.GetName(folderHeader.Hash) ?? folderHeader.Hash.toString(16).toUpperCase().padStart(16, '0');

            const folder = new TrpakFolder();
            folder.Path = folderName;

            for (let j = 0; j < folderHeader.ContentNumber; j++) {
                const content = this.ReadStructTrpakFolderIndex(data, offset);
                offset += TrpakFolderIndex.Size;

                const fileIndex = content.Index;
                const fileName = hashCache.GetName(content.Hash) ?? content.Hash.toString(16).toUpperCase().padStart(16, '0');

                // Try to resolve the full path from the file hash
                const fullPath = hashCache.GetName(fileHashes[fileIndex]) ?? (folderName + fileName);

                // Auto-register resolved paths
                if (hashCache.GetName(fileHashes[fileIndex]) === null) {
                    hashCache.Add(fileHashes[fileIndex], fullPath);
                }

                const trpakFile = new TrpakFile();
                trpakFile.Name = fileName;
                trpakFile.FullName = fullPath;
                trpakFile.Data = fileIndex < files.length ? files[fileIndex] : Buffer.alloc(0);

                folder.Files.push(trpakFile);
            }

            archive.Folders.push(folder);
        }

        return archive;
    }

    private static ReadStructTrpakHeader(data: Buffer, offset: number): TrpakHeader {
        return {
            magic: data.readBigUInt64LE(offset),
            Version: data.readUInt32LE(offset + 8),
            Relocated: data.readUInt32LE(offset + 12),
            FileNumber: data.readUInt32LE(offset + 16),
            FolderNumber: data.readUInt32LE(offset + 20)
        };
    }

    private static ReadStructTrpakFolderHeader(data: Buffer, offset: number): TrpakFolderHeader {
        return {
            Hash: data.readBigUInt64LE(offset),
            ContentNumber: data.readUInt32LE(offset + 8),
            Reserved: data.readUInt32LE(offset + 12)
        };
    }

    private static ReadStructTrpakFolderIndex(data: Buffer, offset: number): TrpakFolderIndex {
        return {
            Hash: data.readBigUInt64LE(offset),
            Index: data.readUInt32LE(offset + 8),
            Reserved: data.readUInt32LE(offset + 12)
        };
    }

    private static ReadStructTrpakFileHeader(data: Buffer, offset: number): TrpakFileHeader {
        return {
            Level: data.readUInt16LE(offset),
            CompressionType: data.readUInt16LE(offset + 2) as TrpakCompressionType,
            BufferSize: data.readUInt32LE(offset + 4),
            FileSize: data.readUInt32LE(offset + 8),
            Reserved: data.readUInt32LE(offset + 12),
            FilePointer: data.readBigUInt64LE(offset + 16)
        };
    }
}
